### DISCUSSION TITLE
Does Tart support nested virtualization?

### DISCUSSION THREAD

**Discussion by jonathandieu** [cirruslabs/tart#701] [Q&A] [answered]
I'd like to be able to run Docker Desktop when running workflows with these runners- is that something that's possible using Tart / Cirrus runners?

Thanks in advance :shipit:

**Comment by fkorotkov** [comment, marked as answer]
Unfortunately, it's not supported at the moment by the `Virtualization.Framework` that Tart depends on even though M1/M2/M3 chips have the hardware capabilities.

**Reply by jonathandieu** [reply]
Thanks for the speedy response! 🙇

**Reply by rpendleton** [reply]
Looks like `Virtualization.framework` might be adding support for nested virtualization starting in macOS 15:

https://developer.apple.com/documentation/virtualization/vzgenericplatformconfiguration/4360552-isnestedvirtualizationenabled?changes=latest_minor

**Reply by fkorotkov** [reply]
Unfortunately only for Linux VMs. Nothing for macOS 😢

**Reply by sorenbs** [reply]
Does Tart support running a Linux guest with nested virtualization?

**Reply by edigaryev** [reply]
Please see https://github.com/cirruslabs/tart/pull/933.

**Comment by fkorotkov** [comment]
As of version `2.20.0` you can enable nested virtualization for Linux VMs by passing `--nested` flag to `tart run`.

**Reply by sorenbs** [reply]
This is great news!

Are you aware of any benchmarks of the overhead of nested virtualization on Apple silicon?

**Reply by nohajc** [reply]
I couldn't find any information as to why nested virtualization doesn't work with macOS guests. Is this a limitation on Apple's side?

If so, is there any chance macOS 26 Tahoe will lift it?

**Reply by fkorotkov** [reply]
Unfortunately, Tahoe didn't change it. Even Linux VMs inside macOS VMs will be super beneficial.

**Reply by nohajc** [reply]
Of course. I just have a very specific use case - developing a project which uses a Linux microVM on macOS. I can automate builds on multiple OS versions using tart but can't run any tests which would involve starting a nested Linux VM by design.

So if I understand it correctly, this is something Apple needs to address, right? Is it even documented anywhere or it's just trial and error on our part?

Thanks!

**Reply by fkorotkov** [reply]
This use case should be supported on M3/M4 with `tart run --nested`. It's only not possible inside of macOS guests and on M1/M2.

**Reply by nohajc** [reply]
Yeah, unfortunately I need the macOS guest with nested Linux. Let's hope Apple enables that soon.

**Reply by nohajc** [reply]
I meant to ask whether Apple documents the fact it doesn't support nested virtualization for macOS guests.

I think tart's documentation is OK. :)

**Reply by paolomainardi** [reply]
If I've understood right, it is only possible to run Linux VM with nested virtualization on, right ?

```shell
❯  tart --version
2.28.5

❯ tart run --nested tahoe
macOS virtual machines do not support nested virtualization

❯ sysctl -a | grep machdep.cpu
machdep.cpu.cores_per_package: 12
machdep.cpu.core_count: 12
machdep.cpu.logical_per_package: 12
machdep.cpu.thread_count: 12
machdep.cpu.brand_string: Apple M4 Pro
```

**Reply by fkorotkov** [reply]
Only Linux VM unfortunately. 😢

**Reply by paolomainardi** [reply]
> Only Linux VM unfortunately. 😢 

Thanks for your quick reply. Can I suggest to update the documentation to specify this ? I guess I can open a PR.

**Reply by fkorotkov** [reply]
@codex maybe you can take care of a PR to update error message and mention that only Linux VMs on M3+ support nested virtualization.
